packages/libsql/src/lib/Category.ts

```diff
@@ -0,0 +1,417 @@
+
+type Decode<E> = (v: ITimelineEvent<Format>) => E | undefined
+
+export class LibSqlConnection {
+  constructor(
+    public read: LibSqlReader,
+    public write: LibSqlWriter,
+  ) {}
+
+  static create(writeClient: Client, readClient = writeClient) {
+    return new LibSqlConnection(new LibSqlReader(readClient), new LibSqlWriter(writeClient))
+  }
+}
+
+type ContextConfig = {
+  client: Client
+  readClient?: Client
+  batchSize: number
+  maxBatches?: number
+}
+
+export class LibSqlContext {
+  constructor(
+    private readonly conn: LibSqlConnection,
+    public readonly batchSize: number,
+    public readonly maxBatches?: number,
+  ) {}
+
+  tokenEmpty = Token.create(0n)
+
+  async loadBatched<Event, State>(
+    streamName: string,
+    decode: Decode<Event>,
+    fold: (state: State, events: Event[]) => State,
+    initial: State,
+  ): Promise<[StreamToken, State]> {
+    let state = initial
+    let version = 0n
+    let batches = 0
+    let eventCount = 0
+    for await (const [lastVersion, events] of Read.loadForwardsFrom(
+      this.conn.read,
+      this.batchSize,
+      this.maxBatches,
+      streamName,
+      0n,
+    )) {
+      batches++
+      eventCount += events.length
+      state = fold(state, keepMap(events, decode))
+      version = lastVersion
+    }
+    trace.getActiveSpan()?.setAttributes({
+      [Tags.loaded_count]: eventCount,
+      [Tags.batches]: batches,
+      [Tags.read_version]: Number(version),
+    })
+    return [Token.create(version), state]
+  }
+
+  async loadLast<Event, State>(
+    streamName: string,
+    decode: Decode<Event>,
+    fold: (state: State, events: Event[]) => State,
+    initial: State,
+  ): Promise<[StreamToken, State]> {
+    const [version, events] = await Read.loadLastEvent(this.conn.read, streamName)
+    trace.getActiveSpan()?.setAttributes({
+      [Tags.loaded_count]: 1,
+      [Tags.read_version]: Number(version),
+    })
+    return [Token.create(version), fold(initial, keepMap(events, decode))]
+  }
+
+  async loadState<Event, State>(
+    streamName: string,
+    decode: Decode<Event>,
+    fold: (state: State, events: Event[]) => State,
+    initial: State,
+  ): Promise<[StreamToken, State]> {
+    const snapshot = await this.conn.read.readSnapshot(streamName)
+    // An end-user might add RollingState to a category after the fact, we should handle this gracefully
+    if (snapshot == null) return this.loadBatched(streamName, decode, fold, initial)
+    const [etag, event] = snapshot
+    const decoded = decode(event)
+    // An end-user might add RollingState to a category after the fact, we should handle this gracefully
+    if (!decoded) return this.loadBatched(streamName, decode, fold, initial)
+    const version = BigInt(event.index)
+    const state = fold(initial, [decoded])
+    const span = trace.getActiveSpan()
+    span?.setAttributes({
+      [Tags.loaded_count]: 1,
+      [Tags.snapshot_version]: Number(version),
```

Comment:
hm back to prev discussion re version vs position vs index on snapshot
in cosmos, I have i which means "this is a snapshot taken when n was this value (n is the version and is the _n_ext index/i")
that would be valid as a version too (i.e. a snapshot with i=1 is indeed a snapshot at version 1)

I _think_ it all makes sense the way you have it, i.e. `index` on events is 0 based, and index on a snapshot is its logical index in the order (comes after the events, last event index+1)
that makes index on a snapshot be equivalent to a version.

OK so most of my other ramblings can prob be disregarded!